### DISCUSSION TITLE
No client-only APIs or next/headers in client component

### DISCUSSION THREAD

**Discussion by tinleym** [vercel/next.js#42893] [Help]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

    Operating System:
      Platform: darwin
      Arch: x64
      Version: Darwin Kernel Version 21.6.0: Mon Aug 22 20:17:10 PDT 2022; root:xnu-8020.140.49~2/RELEASE_X86_64
    Binaries:
      Node: 16.13.1
      npm: 8.1.2
      Yarn: 1.22.15
      pnpm: 6.11.0
    Relevant packages:
      next: 13.0.3-canary.4
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0

### What browser are you using? (if relevant)

Chrome 107.0.5304.110

### How are you deploying your application? (if relevant)

next dev

### Describe the Bug

Client components can render in an environment where neither client-only APIs (like the document object) nor next/headers are available. 

This creates issues when, for example, cookies are needed for a fetch. If a client environment is assumed, the document object isn't available and credentials are not available to be attached to a request. If a server environment is assumed, cookies aren't available from next/headers and so also can't be attached to the request.

### Expected Behavior

If client-only APIs are not always available to client components, then the next/headers API should be available.

### Link to reproduction - Issues with a link to complete (but minimal) reproduction code will be addressed faster

https://github.com/tinleym/next13-oddity

### To Reproduce

Run next dev and check the server-side console. You'll see a case where neither the document object nor the next/headers cookie function is available.

**Comment by balazsorban44** [comment]
There is no way to read headers in a browser. For cookies, you can read them from `document.cookie`, although keep in mind, only non `HttpOnly` cookies can be read.

Cookies are automatically attached to a client-side request to the backend, so you don't need to do anything about that.

To initiate a `fetch` call from a client component, usually you would use a library (`swr` or `react-query`), or `useEffect`, which only executes client-side, and will have `document` available.

**Reply by tinleym** [reply]
Thank you. What I was trying to demonstrate in the repo I linked is that there's an environment where neither headers nor document is available. Not demonstrated, but a fetch in this environment also does not include credentials.

**Comment by kevva** [comment]
@balazsorban44, I can reproduce this as well in `v13.0.5`. For some reason I get a lot of `ReferenceError: document is not defined` errors inside client components that imports modules which references `document`. This can be reproduced by simply creating a `app/page.jsx` file with the following content:

```js
'use client'
import React from 'react';

console.log(document);

export default function Page() {
	return <div>Hello World!</div>;
}
```

However, if I use the `experimental-edge` runtime for that page it seems to work for some reason.

**Reply by balazsorban44** [reply]
Hi, that is expected. You cannot access Web APIs safely outside `useEffect`, as they won't be available during the prerendering pass at build time.

**Reply by icyJoseph** [reply]
Unfortunately, "client components", and a bunch of social media tech content, is making it sound like these components run exclusively on the browser :/ - don't forget about the SSR pre-rendering step!

**Reply by kevva** [reply]
@balazsorban44, I figured. So what is the preferred way of dealing with third party dependencies making use of these APIs? Dynamically import them inside `useEffect`/`useState`? That throws another error:

```
TypeError: wakeable.then is not a function
```

**Reply by icyJoseph** [reply]
You can just hide them behind a flag, ala `useLayoutEffect` documented here https://gist.github.com/gaearon/e7d97cdf38a2907924ea12e4ebdf3c85, simply load them with a flag that is set in the client. 

Though there might still be an issue with side-effects, or top level module code that uses unguarded `window` or DOM access.

Then you can just use a simple import:

```tsx
"use client";

import { useEffect, useState } from "react"

export const LoadModule = () => {
  const [mod, setMod] = useState<Awaited<
    typeof import("../utils/cookie")
  > | null>(null);

  useEffect(() => {
    import("../utils/cookie").then(setMod);
  }, []);

  if (!mod) return <div>Loading module</div>;

  console.log("cookies", mod.readCookies());

  return <div>Module loaded</div>;
};

```

Where `utils` is:

```ts
console.log(typeof window); // won't crash even if this is used as a top level side-effect-ish
export const readCookies = () => document.cookie;
```

**Comment by urwrstkn8mare** [comment]
Yeah, this is an issue, especially when needing to access cookies. I've had to make some unsavoury compromises because of this, so if this could be resolved, that would be great. IMO, this could be resolved by allowing 2 imports depending on if the component is being rendered on the server or client (e.g. if server than it will import `util.server.ts` and if it's on the client than `util.client.ts`, that way different logic can be used which is necessary because of the different ways cookies are retrieved).
